File: extensions/jetbrains/src/main/kotlin/com/aistack/client/AistackClient.kt
package com.aistack.client

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Minimal REST client for the aistack daemon. Mirrors the VS Code extension's
 * surface: list/spawn/stop agents, list memory entries, list/start review loops.
 *
 * Uses OkHttp + kotlinx.serialization. Single retry on 5xx / network failure.
 * The daemon is local so we keep timeouts tight.
 */
class AistackClient(
    private val baseUrl: String,
    private val token: String?,
    private val timeoutMs: Long
) {
    private val http: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(timeoutMs, TimeUnit.MILLISECONDS)
        .readTimeout(timeoutMs, TimeUnit.MILLISECONDS)
        .writeTimeout(timeoutMs, TimeUnit.MILLISECONDS)
        .build()

    private val json: Json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
    }

    private val jsonMedia = "application/json".toMediaType()

    // ---------- data classes ----------

    @Serializable
    data class AgentSummary(
        val id: String,
        val type: String,
        val name: String? = null,
        val status: String,
        val sessionId: String? = null,
        val createdAt: String? = null
    )

    @Serializable
    data class AgentTypeDef(
        val type: String,
        val name: String,
        val description: String,
        val capabilities: List<String> = emptyList()
    )

    @Serializable
    data class SpawnAgentRequest(
        val type: String,
        val name: String? = null,
        val sessionId: String? = null
    )

    @Serializable
    data class ExecuteAgentRequest(val task: String)

    @Serializable
    data class ExecuteAgentResponse(
        val response: String,
        val model: String? = null,
        val duration: Long? = null
    )

    @Serializable
    data class ReviewLoopSummary(
        val id: String,
        val coderId: String,
        val adversarialId: String,
        val iteration: Int,
        val maxIterations: Int,
        val status: String,
        val startedAt: String? = null,
        val completedAt: String? = null,
        val reviewCount: Int = 0
    )

    @Serializable
    data class StartReviewLoopRequest(
        val codeInput: String,
        val maxIterations: Int? = null
    )

    // ---------- agents ----------

    fun listAgents(): List<AgentSummary> =
        getJson("/api/v1/agents", kotlinx.serialization.builtins.ListSerializer(AgentSummary.serializer()))

    fun listAgentTypes(): List<AgentTypeDef> =
        getJson(
            "/api/v1/agents/types",
            kotlinx.serialization.builtins.ListSerializer(AgentTypeDef.serializer())
        )

    fun spawnAgent(req: SpawnAgentRequest): AgentSummary =
        postJson("/api/v1/agents", req, SpawnAgentRequest.serializer(), AgentSummary.serializer())

    fun stopAgent(id: String) {
        val request = baseRequest("/api/v1/agents/${encode(id)}").delete().build()
        execute(request).close()
    }

    fun executeAgent(id: String, task: String): ExecuteAgentResponse =
        postJson(
            "/api/v1/agents/${encode(id)}/execute",
            ExecuteAgentRequest(task),
            ExecuteAgentRequest.serializer(),
            ExecuteAgentResponse.serializer()
        )

    // ---------- review loops ----------

    fun listReviewLoops(): List<ReviewLoopSummary> =
        getJson(
            "/api/v1/review-loops",
            kotlinx.serialization.builtins.ListSerializer(ReviewLoopSummary.serializer())
        )

    fun startReviewLoop(req: StartReviewLoopRequest): ReviewLoopSummary =
        postJson(
            "/api/v1/review-loops",
            req,
            StartReviewLoopRequest.serializer(),
            ReviewLoopSummary.serializer()
        )

    // ---------- internals ----------

    private fun <T> getJson(path: String, deserializer: kotlinx.serialization.KSerializer<T>): T {
        val req = baseRequest(path).get().build()
        val body = execute(req).use { it.body?.string().orEmpty() }
        return json.decodeFromString(deserializer, body)
    }

    private fun <Req, Resp> postJson(
        path: String,
        body: Req,
        reqSer: kotlinx.serialization.KSerializer<Req>,
        respSer: kotlinx.serialization.KSerializer<Resp>
    ): Resp {
        val payload = json.encodeToString(reqSer, body).toRequestBody(jsonMedia)
        val request = baseRequest(path).post(payload).build()
        val responseBody = execute(request).use { it.body?.string().orEmpty() }
        return json.decodeFromString(respSer, responseBody)
    }

    private fun baseRequest(path: String): Request.Builder {
        val b = Request.Builder().url(baseUrl.trimEnd('/') + path).header("Accept", "application/json")
        if (!token.isNullOrBlank()) b.header("Authorization", "Bearer $token")
        return b
    }

    private fun execute(request: Request, attempt: Int = 0): okhttp3.Response {
        val resp: okhttp3.Response = try {
            http.newCall(request).execute()
        } catch (e: IOException) {
            if (attempt == 0) {
                Thread.sleep(500)
                return execute(request, attempt + 1)
            }
            throw AistackException(
                -1,
                "aistack daemon unreachable at $baseUrl: ${e.message}"
            )
        }
        if (!resp.isSuccessful) {
            val status = resp.code
            val text = resp.body?.string().orEmpty()
            resp.close()
            if (status >= 500 && attempt == 0) {
                Thread.sleep(500)
                return execute(request, attempt + 1)
            }
            throw AistackException(status, "${request.method} ${request.url.encodedPath} failed: $status $text")
        }
        return resp
    }

    private fun encode(s: String): String =
        java.net.URLEncoder.encode(s, Charsets.UTF_8).replace("+", "%20")
}

class AistackException(val status: Int, message: String) : RuntimeException(message)

File: extensions/vscode/src/client/aistackClient.ts
/**
 * REST client for the aistack daemon.
 *
 * Uses the global `fetch` (Node 20+ / VS Code's runtime). Provides typed
 * helpers for the small subset of the daemon API this extension consumes:
 *
 *   - GET    /api/v1/agents
 *   - GET    /api/v1/agents/types
 *   - POST   /api/v1/agents
 *   - DELETE /api/v1/agents/:id
 *   - POST   /api/v1/agents/:id/execute
 *   - GET    /api/v1/memory
 *   - GET    /api/v1/review-loops
 *   - POST   /api/v1/review-loops
 *
 * Retries: 1 retry with 500ms backoff on network / 5xx errors. The daemon is
 * expected to be on localhost so aggressive retry is unwarranted.
 */

import * as vscode from 'vscode';

export interface ClientConfig {
  baseUrl: string;
  token?: string;
  timeoutMs: number;
}

export interface AgentSummary {
  id: string;
  type: string;
  name?: string;
  status: 'idle' | 'running' | 'completed' | 'failed' | 'stopped';
  sessionId?: string;
  createdAt: string;
}

export interface AgentTypeDefinition {
  type: string;
  name: string;
  description: string;
  capabilities: string[];
}

export interface MemoryEntry {
  id: string;
  namespace?: string;
  key: string;
  value: unknown;
  createdAt: string;
}

export interface ReviewLoopSummary {
  id: string;
  coderId: string;
  adversarialId: string;
  sessionId?: string;
  iteration: number;
  maxIterations: number;
  status: 'running' | 'approved' | 'aborted' | 'max_iterations_reached';
  finalVerdict?: string;
  startedAt: string;
  completedAt?: string;
  reviewCount: number;
}

export class AistackHttpError extends Error {
  constructor(public readonly status: number, message: string) {
    super(message);
    this.name = 'AistackHttpError';
  }
}

export class AistackClient {
  constructor(private readonly cfg: ClientConfig, private readonly output: vscode.OutputChannel) {}

  get baseUrl(): string {
    return this.cfg.baseUrl.replace(/\/+$/, '');
  }

  // ---------- agents ----------

  listAgents(): Promise<AgentSummary[]> {
    return this.req<AgentSummary[]>('GET', '/api/v1/agents');
  }

  listAgentTypes(): Promise<AgentTypeDefinition[]> {
    return this.req<AgentTypeDefinition[]>('GET', '/api/v1/agents/types');
  }

  spawnAgent(input: { type: string; name?: string; sessionId?: string }): Promise<AgentSummary> {
    return this.req<AgentSummary>('POST', '/api/v1/agents', input);
  }

  stopAgent(id: string): Promise<{ stopped: boolean }> {
    return this.req('DELETE', `/api/v1/agents/${encodeURIComponent(id)}`);
  }

  executeAgent(
    id: string,
    task: string,
    extras?: { provider?: string; model?: string; context?: Record<string, unknown> }
  ): Promise<{ response: string; model?: string; duration?: number }> {
    return this.req('POST', `/api/v1/agents/${encodeURIComponent(id)}/execute`, {
      task,
      ...extras,
    });
  }

  // ---------- memory ----------

  listMemory(opts?: { sessionId?: string; limit?: number; offset?: number }): Promise<{
    items: MemoryEntry[];
    total: number;
  }> {
    const qs = new URLSearchParams();
    if (opts?.sessionId) qs.set('sessionId', opts.sessionId);
    if (opts?.limit != null) qs.set('limit', String(opts.limit));
    if (opts?.offset != null) qs.set('offset', String(opts.offset));
    const suffix = qs.toString() ? `?${qs.toString()}` : '';
    return this.req('GET', `/api/v1/memory${suffix}`);
  }

  // ---------- review loops ----------

  listReviewLoops(): Promise<ReviewLoopSummary[]> {
    return this.req('GET', '/api/v1/review-loops');
  }

  startReviewLoop(input: {
    codeInput: string;
    maxIterations?: number;
    sessionId?: string;
  }): Promise<ReviewLoopSummary> {
    return this.req('POST', '/api/v1/review-loops', input);
  }

  // ---------- internals ----------

  private async req<T>(method: string, path: string, body?: unknown, attempt = 0): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.cfg.timeoutMs);
    const headers: Record<string, string> = { Accept: 'application/json' };
    if (body !== undefined) headers['Content-Type'] = 'application/json';
    if (this.cfg.token) headers['Authorization'] = `Bearer ${this.cfg.token}`;

    try {
      const res = await fetch(url, {
        method,
        headers,
        body: body !== undefined ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });
      if (!res.ok) {
        const text = await res.text().catch(() => '');
        // Retry once on 5xx
        if (res.status >= 500 && attempt === 0) {
          this.output.appendLine(`aistack: ${method} ${path} -> ${res.status}, retrying once`);
          await delay(500);
          return this.req<T>(method, path, body, attempt + 1);
        }
        throw new AistackHttpError(res.status, `${method} ${path} failed: ${res.status} ${text}`);
      }
      if (res.status === 204) return undefined as T;
      return (await res.json()) as T;
    } catch (err) {
      if (err instanceof AistackHttpError) throw err;
      // Network / abort — retry once
      if (attempt === 0) {
        this.output.appendLine(`aistack: ${method} ${path} network error, retrying once`);
        await delay(500);
        return this.req<T>(method, path, body, attempt + 1);
      }
      const message = err instanceof Error ? err.message : String(err);
      throw new Error(`aistack daemon unreachable at ${this.baseUrl}: ${message}`);
    } finally {
      clearTimeout(timer);
    }
  }
}

function delay(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}
